src/planner/table_binding.cpp

```diff
@@ -12,39 +12,53 @@
 
 namespace duckdb {
 
-Binding::Binding(const string &alias, vector<LogicalType> coltypes, vector<string> colnames, idx_t index)
-    : alias(alias), index(index), types(move(coltypes)), names(move(colnames)) {
+Binding::Binding(const string &alias, vector<LogicalType> coltypes, vector<string> colnames,
+                 vector<LogicalType> gcoltypes, vector<string> gcolnames, idx_t index)
+    : alias(alias), index(index), types(move(coltypes)), names(move(colnames)), gtypes(move(gcoltypes)),
+      gnames(move(gcolnames)) {
 	D_ASSERT(types.size() == names.size());
+	D_ASSERT(gtypes.size() == gnames.size());
 	for (idx_t i = 0; i < names.size(); i++) {
```

Comment:
Rather than having all these different vectors as parameters (`coltypes, colnames, gcoltypes, gcolnames`) could we perhaps have one vector with a struct, e.g.:

```cpp
struct BindingColumnInfo {
    string name;
    LogicalType type;
    TableColumnType column_type;
};
```

That should simplify the code here.

Reply:
`// TODO: bundle the type and name vector in a struct (e.g PackedColumnMetadata)`

I had a similar idea, though mine would still involve passing 2 variables instead of just 1 - so I'm glad I waited because this is better
I'll make the change now

Reply:
Working on this, it's looking like it introduces quite a couple extra loops where the columns are used, any place where names.size() is used, maybe the struct containing the two vectors and a type might be better

Reply:
```cpp
void BindContext::AddSubquery(idx_t index, const string &alias, SubqueryRef &ref, BoundQueryNode &subquery) {
	auto names = AliasColumnNames(alias, subquery.names, ref.column_name_alias);
	AddGenericBinding(index, alias, names, subquery.types);
}

void BindContext::AddSubquery(idx_t index, const string &alias, TableFunctionRef &ref, BoundQueryNode &subquery) {
	auto names = AliasColumnNames(alias, subquery.names, ref.column_name_alias);
	AddGenericBinding(index, alias, names, subquery.types);
}
```
things like this make me hesitant to make this change, because I think we also want to add the option to have generated columns to the GenericBinding (the expression validity check is done using a GenericBinding, so I believe if we want to support generated_columns in generated expression references we are going to have to do this)

Reply:
Ended up just having a optional `vector<TableColumnType>` overload of the AddGenericBinding, used when populating the name_map